### ISSUE TITLE
[test-framework] Add Kube Environment Support for MCPServer component.

### ISSUE THREAD

**Issue by ozevren** [istio/istio#13466]
Similar to the PolicyBackend, MCP Server should be remotely deployable and controllable.

**Comment by su225** [comment]
I'm new to the Istio project and interested in contributing. Can I take this one?

**Comment by ozevren** [comment]
Sure.
The current MCP server facade here only supports running in native mode: https://github.com/istio/istio/tree/master/pkg/test/framework/components/mcpserver

We'd like this to have a K8s counterpart, where we it gets deployed as a binary and acts as a remote agent that can be controlled by the tests.

The PolicyBackend is an example of this, that has two gRPC services, one serving Out-of-process adapter protocol (for Mixer integration), and the other for controlling its behavior from with in tests.
https://github.com/istio/istio/tree/master/pkg/test/framework/components/policybackend

The MCP Server should work in a similar way.

**Comment by ozevren** [comment]
I'll assign myself to take it off the table. You can work on it.

**Comment by su225** [comment]
Here is my understanding so far. 

1.In-Process for testing MCP server directly where as the Out-of-process is for integration with Galley (dialout_test can then run in Kubernetes environment. So I can test whether my changes work by enabling it to run on Kubernetes environment and checking if this test goes fine).
2. Need a fake MCP server running in sink mode (something like the one in istio/mixer/test/policybackend/main.go). I see that there is one in **istio/pkg/mcp/testing/server.go**. But it starts the server in Source mode. That must be modified to make the mode configurable. Docker image can then be created for the fake based on that.
3. Need protobuf definition like istio/mixer/test/keyval/template_handler_service.proto and the corresponding server and client for controlling the test instance remotely.
4. Finally adding k8s support for the mcpserver component.

**Comment by ozevren** [comment]
You're pretty much on point. You don't need to write a new service based on istio/pkg/mcp/testing/server.go. You can take the native code in framework/components/mcpserver and enhance it.

in-process v.s. out-process is a Mixer specific thing, about how Mixer adapters work. You don't need it for MCP.

**Comment by su225** [comment]
I have opened a work-in-progress pull request. I need some help to integrate my changes so that I can run galley requests.

**Comment by su225** [comment]
I added Make targets and was able to use run the Galley test. I followed the pattern in policybackend with some changes. But it keeps giving me the following error telling me that I cannot connect to the MCP Sink through forwarder.

1. I tried exposing the port in Dockerfile.
2. Double-checked forwarder address - it is pointing to 127.0.0.1:<randomly-chosen-port> to the port used in the service (that is 6666). I tried to bring up Istio and then running the test_mcpserver pod manually and I was able to hit it with port-forwarding. But not in the test. Is the error something else? Pod logs indicate closing transport.

```
--- FAIL: TestDialout_Basic (43.11s)
    dialout_test.go:56: Applying configuration: 
        apiVersion: networking.istio.io/v1alpha3
        kind: Gateway
        metadata:
          name: helloworld-gateway
        spec:
          selector:
            istio: ingressgateway # use istio default controller
          servers:
          - port:
              number: 80
              name: http
              protocol: HTTP
            hosts:
            - "*"
    dialout_test.go:58: retry.UntilSuccessOrFail: timeout while waiting (last error: expected number of objects not found (current: 0))
2019-05-27T18:38:33.282305Z	info	(0xc0004a6210) (0xc00066c0a0) Stream removed, broadcasting: 3
2019-05-27T18:38:33.282342Z	info	transport: loopyWriter.run returning. connection error: desc = "transport is closing"
2019-05-27T18:38:33.282363Z	error	error closing listener: close tcp4 127.0.0.1:44047: use of closed network connection
2019-05-27T18:38:33.282386Z	error	error closing listener: close tcp6 [::1]:44047: use of closed network connection
2019-05-27T18:38:33.282476Z	info	frame read error: read tcp 192.168.39.1:48454->192.168.39.89:8443: use of closed network connection
2019-05-27T18:38:33.282485Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc00020b470, TRANSIENT_FAILURE
2019-05-27T18:38:33.282502Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc00020b470, CONNECTING
2019-05-27T18:38:33.282560Z	info	grpc: addrConn.createTransport failed to connect to {127.0.0.1:44047 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:44047: connect: connection refused". Reconnecting...
```

**Comment by ozevren** [comment]
Hard to tell from the logs what is going on. None of the ports on the logs match what you've mentioned. It is possible that the service maybe listening on 127.0.0.1 instead of 0.0.0.0 and not responding. Another possibility is the port mismatch on the Galley side. The port-forwarder should auto-allocate the port, which needs to be passed to Galley for establishing the connection.

**Comment by su225** [comment]
No. It is not 127.0.0.1 problem. I had changed it to 0.0.0.0 and still the issue persisted. 127.0.0.1:44047 is the port-forwarder address which is used to contact the MCP Sink. I added more logging and I can see that port-forwarder can start without errors and the connection to MCP sink is also established successfully. 
```
2019-05-30T20:04:56.006948Z	info	tf	forwading from 127.0.0.1:36297 -> 6666
2019-05-30T20:04:56.006962Z	info	tf	client: trying to dial 127.0.0.1:36297 to connect to mcpserver
2019-05-30T20:04:56.006973Z	info	parsed scheme: ""
2019-05-30T20:04:56.006980Z	info	scheme "" not registered, fallback to default scheme
2019-05-30T20:04:56.006993Z	info	ccResolverWrapper: sending update to cc: {[{127.0.0.1:36297 0  <nil>}] }
2019-05-30T20:04:56.006998Z	info	ClientConn switching balancer to "pick_first"
2019-05-30T20:04:56.007039Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc000467980, CONNECTING
2019-05-30T20:04:56.007220Z	info	(0xc00037a9a0) (0xc0005e6000) Create stream
2019-05-30T20:04:56.007269Z	info	(0xc00037a9a0) (0xc0005e6000) Stream added, broadcasting: 1
2019-05-30T20:04:56.008811Z	info	(0xc00037a9a0) Reply frame received for 1
2019-05-30T20:04:56.008837Z	info	(0xc0005e6000) (1) Writing data frame
2019-05-30T20:04:56.008939Z	info	(0xc00037a9a0) (0xc0005e60a0) Create stream
2019-05-30T20:04:56.008952Z	info	(0xc00037a9a0) (0xc0005e60a0) Stream added, broadcasting: 3
2019-05-30T20:04:56.009450Z	info	(0xc00037a9a0) Reply frame received for 3
2019-05-30T20:04:56.009482Z	info	(0xc0005e60a0) (3) Writing data frame
2019-05-30T20:04:56.012957Z	info	(0xc00037a9a0) Data frame received for 3
2019-05-30T20:04:56.012974Z	info	(0xc0005e60a0) (3) Data frame handling
2019-05-30T20:04:56.012980Z	info	(0xc0005e60a0) (3) Data frame sent
2019-05-30T20:04:56.012983Z	info	(0xc00037a9a0) Data frame received for 3
2019-05-30T20:04:56.012985Z	info	(0xc0005e60a0) (3) Data frame handling
2019-05-30T20:04:56.013026Z	info	(0xc0005e60a0) (3) Data frame sent
2019-05-30T20:04:56.013048Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc000467980, READY
2019-05-30T20:04:56.013058Z	info	tf	established connection with mcp-sinkserver at 127.0.0.1:36297
2019-05-30T20:04:56.013062Z	info	(0xc0005e60a0) (3) Writing data frame
2019-05-30T20:04:56.040420Z	info	parsed scheme: ""
2019-05-30T20:04:56.040436Z	info	scheme "" not registered, fallback to default scheme
2019-05-30T20:04:56.040446Z	info	ccResolverWrapper: sending update to cc: {[{127.0.0.1:44341 0  <nil>}] }
2019-05-30T20:04:56.040451Z	info	ClientConn switching balancer to "pick_first"
2019-05-30T20:04:57.111211Z	info	tf	sending GetCollectionState request: collection:"istio/networking/v1alpha3/gateways" 
2019-05-30T20:04:57.111364Z	info	(0xc0005e60a0) (3) Writing data frame
2019-05-30T20:04:57.116764Z	info	(0xc00037a9a0) Data frame received for 3
2019-05-30T20:04:57.116780Z	info	(0xc0005e60a0) (3) Data frame handling
2019-05-30T20:04:57.116785Z	info	(0xc0005e60a0) (3) Data frame sent
2019-05-30T20:04:57.116918Z	info	(0xc0005e60a0) (3) Writing data frame
2019-05-30T20:04:57.116978Z	info	tf	response for collection istio/networking/v1alpha3/gateways is 
2019-05-30T20:04:57.117654Z	info	(0xc00037a9a0) Data frame received for 3
2019-05-30T20:04:57.117666Z	info	(0xc0005e60a0) (3) Data frame handling
2019-05-30T20:04:57.117673Z	info	(0xc0005e60a0) (3) Data frame sent
2019-05-30T20:04:57.127181Z	info	tf	sending GetCollectionState request: collection:"istio/networking/v1alpha3/gateways" 
2019-05-30T20:04:57.127327Z	info	(0xc0005e60a0) (3) Writing data frame
2019-05-30T20:04:57.128484Z	info	(0xc00037a9a0) Data frame received for 3
2019-05-30T20:04:57.128513Z	info	(0xc0005e60a0) (3) Data frame handling
2019-05-30T20:04:57.128517Z	info	(0xc0005e60a0) (3) Data frame sent
2019-05-30T20:04:57.128625Z	info	tf	response for collection istio/networking/v1alpha3/gateways is 
2019-05-30T20:04:57.128662Z	info	(0xc0005e60a0) (3) Writing data frame
2019-05-30T20:04:57.129209Z	info	(0xc00037a9a0) Data frame received for 3
2019-05-30T20:04:57.129217Z	info	(0xc0005e60a0) (3) Data frame handling
2019-05-30T20:04:57.129223Z	info	(0xc0005e60a0) (3) Data frame sent

...... The last few lines repeat many times before timing out.
```

I tried searching for `scheme not registered` issue. But no luck there. Is there any issue in the definition of protobuf (I don't see any) and generating code with protoc (`protoc --version` gives me libprotoc 3.6.0). I generated them with the protoc installed in my machine. 

I see that the fileDescriptor was missing from the generated code earlier. Even after I managed to get it - still no luck. After adding some log lines I see that when I call the required method I get all those log lines ike "Writing Data frame", "Data frame received for 3" and so on.

Exact command I used from Istio root directory
```
protoc pkg/test/fakes/mcpserver/controller.proto --go_out=plugins=grpc:. -I. -I$GOPATH/src/istio.io/istio/vendor/istio.io/api -I$GOPATH/src/istio.io/istio/vendor/github.com/gogo/protobuf
```

**Comment by su225** [comment]
I have one more question. How does galley know the address of mcp-server in kubernetes environment? I don't see cfg being used anywhere other than setting `cfg` field being set in `kubeComponent`. But it is used to pass sink address in case of native? 

I digged into Galley's server mode code and Dockerfile.galley and nowhere I found the sinkAddress command-line argument passed to it. How will Galley know the sink address to dial to without that argument?

**Comment by su225** [comment]
When I changed it to 0.0.0.0, the issue was still there. The address in the logs is the address of the port-forwarder. The test is trying to dial to MCP server through it. I see that Galley uses its own port-forwarder. But what galley needs is **`sinkAddress`** which is not supplied.

One issue. I see that in galley component for k8s (istio/pkg/test/framework/components/galley/kube.go) we just get a handle to existing pod. But I see that in order for Galley to work like it does for native test, it must set `sinkAddress`. I don't see it set in case of K8s. Only when SinkAddress (**here it must be the service address of MCP Sink server like mcp-server.<ns>.svc.cluster.local since both Galley and MCP Sink live inside K8s**) is set Galley will try to contact MCP Sink Server (istio/galley/pkg/server/callout.go). 

It seems that simply passing `SinkAddress` in dialout_test doesn't work. The Galley pod running in K8s must become aware of MCP Server at runtime and update `sinkAddress` accordingly. Or a _[very hacky]_ solution would be to shoot down the Galley deployment, create a similar pod with Sink address available for Galley and redeploy.

**Comment by su225** [comment]
Now I'm really stuck. Native works because we first create MCP Sink and then pass that address to the Galley which comes next. But in K8s environment, Galley will be running without `sinkAddress` (and `sinkAuth` if needed) before MCP sink is brought up. As far as I know `sinkAddress` and `sinkAuth` must be passed at startup to Galley to create `callout`.

To confirm this, I disabled creating istio environment before test and created the environment manually through `install/kubernetes/istio-init.yaml` and `install/kubernetes/istio-mcp.yaml`. In the latter YAML file I added `--sinkAddress=mcp-sinkserver.mcp.svc.cluster.local:6666 --sinkAuth=NONE` to the arguments list in Galley deployment. In the test, I fixed the namespace to `mcp` and brought up MCP Sink. From the logs of Galley I can see that connection is established successfully to my sink with Gateway as the supported resource. 

Galley logs
```
2019-06-01T08:43:23.781054Z     info    server  Starting MCP Source Client connection to: mcp-sinkserver.mcp.svc.cluster.local:6666
2019-06-01T08:43:23.781213Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc0006cb9e0, CONNECTING
2019-06-01T08:43:23.781589Z     info    mcp     (re)trying to establish new MCP source stream
2019-06-01T08:43:23.925321Z     warn    validation      istio-system/istio-galley endpoint not ready: no ready addresses
2019-06-01T08:43:24.840866Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc0006cb9e0, READY
2019-06-01T08:43:24.841662Z     info    mcp     New MCP source stream created
2019-06-01T08:43:24.842748Z     info    mcp     MCP: connection {addr=10.101.223.24:6666 id=1}: NEW (ResourceSource), supported collections: []string{"istio/mixer/v1/config/client/quotaspecbindings", "istio/config/v1alpha2/legacy/signalfx
s", "k8s/core/v1/endpoints", "istio/policy/v1beta1/rules", "istio/config/v1alpha2/legacy/logentries", "istio/config/v1alpha2/legacy/stdios", "istio/config/v1alpha2/legacy/circonuses", "istio/config/v1alpha2/legacy/cloudwatches", "k8s/exte
nsions/v1beta1/ingresses", "istio/networking/v1alpha3/sidecars", "istio/config/v1alpha2/legacy/dogstatsds", "istio/config/v1alpha2/httpapispecs", "istio/config/v1alpha2/legacy/kubernetesenvs", "istio/networking/v1alpha3/gateways", "istio/
policy/v1beta1/handlers", "istio/config/v1alpha2/legacy/metrics", "istio/mixer/v1/config/client/quotaspecs", "istio/config/v1alpha2/templates", "istio/policy/v1beta1/attributemanifests", "istio/networking/v1alpha3/synthetic/serviceentries
", "istio/rbac/v1alpha1/servicerolebindings", "istio/config/v1alpha2/legacy/checknothings", "istio/config/v1alpha2/legacy/kuberneteses", "istio/rbac/v1alpha1/clusterrbacconfigs", "istio/config/v1alpha2/legacy/prometheuses", "istio/config/
v1alpha2/legacy/quotas", "istio/config/v1alpha2/legacy/redisquotas", "istio/config/v1alpha2/legacy/tracespans", "istio/config/v1alpha2/legacy/statsds", "istio/networking/v1alpha3/virtualservices", "istio/config/v1alpha2/legacy/listentries
", "istio/networking/v1alpha3/envoyfilters", "istio/config/v1alpha2/legacy/solarwindses", "istio/config/v1alpha2/legacy/noops", "istio/rbac/v1alpha1/serviceroles", "istio/rbac/v1alpha1/authorizationpolicies", "istio/config/v1alpha2/legacy
/opas", "istio/mesh/v1alpha1/MeshConfig", "istio/config/v1alpha2/legacy/zipkins", "istio/networking/v1alpha3/serviceentries", "istio/config/v1alpha2/legacy/fluentds", "istio/config/v1alpha2/legacy/authorizations", "k8s/core/v1/namespaces"
, "istio/config/v1alpha2/legacy/deniers", "istio/config/v1alpha2/legacy/reportnothings", "istio/authentication/v1alpha1/meshpolicies", "k8s/core/v1/nodes", "istio/rbac/v1alpha1/rbacconfigs", "istio/config/v1alpha2/legacy/memquotas", "k8s/
core/v1/services", "istio/config/v1alpha2/legacy/bypasses", "istio/config/v1alpha2/legacy/rbacs", "istio/authentication/v1alpha1/policies", "istio/config/v1alpha2/legacy/edges", "istio/config/v1alpha2/legacy/apikeys", "istio/config/v1alph
a2/adapters", "istio/policy/v1beta1/instances", "istio/config/v1alpha2/httpapispecbindings", "istio/config/v1alpha2/legacy/stackdrivers", "istio/networking/v1alpha3/destinationrules", "k8s/core/v1/pods", "istio/config/v1alpha2/legacy/list
checkers"}
2019-06-01T08:43:24.845461Z     info    mcp     MCP: connection {addr=10.101.223.24:6666 id=1}: inc=false WATCH for istio/networking/v1alpha3/gateways
2019-06-01T08:43:24.845872Z     info    mcp     Watch(): created watch 1 for istio/networking/v1alpha3/gateways from group "default", version ""
2019-06-01T08:43:26.586172Z     info    runtime Synchronization is complete, starting distribution.
2019-06-01T08:43:27.587378Z     info    mcp     SetSnapshot(): respond to watch 1 for istio/networking/v1alpha3/gateways @ version "0"
2019-06-01T08:43:27.589412Z     info    mcp     MCP: connection {addr=10.101.223.24:6666 id=1} ACK collection=istio/networking/v1alpha3/gateways with version="0" nonce="1" inc=false
```

Galley is sending Gateway information to my sink. Still it wasn't successful though. Sink sometimes sends ACK and sometimes NACK. Here is the log from my fake sink. It receives config from Galley. I just found https://godoc.org/github.com/golang/protobuf/ptypes#Empty to that error message. 
```
2019-06-01T08:43:31.726682Z     error   mcp     mcpserver.sink - sending resources system_version_info:"65" collection:"istio/networking/v1alpha3/gateways" resources:<metadata:<name:"dialout-2-54448/helloworld-gateway" create_time:<seconds:1559378610 > version:"60471" annotations:<key:"kubectl.kubernetes.io/last-applied-configuration" value:"{\"apiVersion\":\"networking.istio.io/v1alpha3\",\"kind\":\"Gateway\",\"metadata\":{\"annotations\":{},\"name\":\"helloworld-gateway\",\"namespace\":\"dialout-2-54448\"},\"spec\":{\"selector\":{\"istio\":\"ingressgateway\"},\"servers\":[{\"hosts\":[\"*\"],\"port\":{\"name\":\"http\",\"number\":80,\"protocol\":\"HTTP\"}}]}}\n" > > body:<type_url:"type.googleapis.com/istio.networking.v1alpha3.Gateway" value:"\n\023\n\016\010P\022\004HTTP\032\004http\022\001*\022\027\n\005istio\022\016ingressgateway" > > nonce:"2" 
2019-06-01T08:43:31.727350Z     error   mcp     MCP: sending NACK for nonce=2: error="any: message type \"istio.networking.v1alpha3.Gateway\" isn't linked in"
2019-06-01T08:43:39.441608Z     error   mcp     mcpserver.sink - sending resources system_version_info:"72" collection:"istio/networking/v1alpha3/gateways" nonce:"3" 
```

Big problem to make Galley and Sink talk to each other. As noted above, **it is not as simple as passing `SinkAddress` parameter**. Possible solutions might be
1. [VERY HACKY] Bring down Galley deployment, add the sinkAddress argument and redeploy again
2. [INELEGANT] Providing some way of passing additional parameters before deploying Galley. Problem is, the exact address of MCP sink is not known and Galley is deployed before that in K8s. However, we can have a separate K8s version of `dialout_test` where we first bring up MCP sink, then bring the rest of Istio system up (with patched configuration for Galley).
3. [OVERKILL] Make sinkAddress piece dynamically configurable. An overkill? Because of additional headache it might bring in when it comes to maintaining `callout`. 

Port forwarder stuff is working fine. That is not an issue.

**Comment by ozevren** [comment]
Can we pass the address of the sink server, using the local DNS based
service name (similar to how Pilot connects to Galley in K8s)? Then, we can
deploy Istio with custom configuration, pointing to the service name. We
should be able to plumb the callout parameter through to Galley in case of
custom deployment.

**Comment by su225** [comment]
Updated the PR. Not in the final form yet, but it works.

> Can we pass the address of the sink server, using the local DNS based service name (similar to how Pilot connects to Galley in K8s)?

Yup. That's what I did while hacking around. Because both MCP sink and Galley live in K8s cluster. That is not the problem.

> We should be able to plumb the callout parameter through to Galley in case of custom deployment.

This is the actual problem. That is why I have to fetch Galley deployment config and change command run to include `sinkAddress` and update it. This is the hack I talked about. Coming to customizing Istio deployment - I see that I can point to custom Helm repository with `SetupConfigFn`. I haven't explored this part yet. For now, things work fine. But still I would call modifying Galley deployment a hack.

**Comment by ozevren** [comment]
I think the right way to do this is in the deployment phase (i.e. ``SetupConfigFn``). We need to reconcile this with the native mode setup, however, as the native mode setup is done through Galley component configuration. In K8s mode, we can error out if the dialout settings are specified in the config. 

This will make the calling code a bit uglier (it has to take environment into account, in a couple of places), but I think it is the right compromise, and protects caller from accidentally writing erroneous tests.

**Comment by su225** [comment]
I think the clean way to do this is as follows
1. Change helm chart to include `sinkAddress` and `sinkAuth` in Galley templates. This can be done safely without affecting other deployments.
2. Once the change is done, we can then pass `SetupConfigFn` to pass Helm values to config.

The code would be environment specific like you said. There are other minor concerns like setup MCP server before Istio and its teardown. This can be reconciled cleanly with native tests.

**Comment by ozevren** [comment]
I don't you can land #1, as this was objected to by Environment WG lead earlier. Given we have the environment variable support in Galley config, we should be able to utilize that.

**Comment by su225** [comment]
Sorry if I'm missing something here. I don't see how I can pass MCP sink address through environment variables in Galley. When I bring up MCP Server before Galley, its address will be in Galley's pod environment. I went through the code, the helm deployment and couldn't find a way where Galley reads environment variable without changing its code.

As a compromise, to achieve the same effect I could apply the Deployment config update hack I did for dialout_test **at the time of setup**. (I have moved this test to another package so that the change in setup function doesn't affect other Galley tests). Something like this (not the real code)
```go
func TestMain(m *testing.M) {
	framework.
		NewSuite("galley_with_callout", m).
		Setup(setupMCPSinkServer).
		SetupOnEnv(environment.Kube, setupIstioOnKubernetes).
                SetupOnEnv(environment.Kube, updateGalleyDeploymentConfigWithSinkAddress).
		Run()
}
```

From the tests - be it native or Kubernetes, we don't create MCP server instance - instead try to obtain a handle to the instance created during setup. 
```go
srv := mcpserver.GetSinkHandleOrFail(t, ctx)
```
This means that the instance must be in test context. For Galley, as you suggested it will be environment specific instead of environment agnostic `galley.NewOrFail`. 

This compromise cleanly reconciles tests in both environments and also safety requirement you mentioned. Setup is a bit involved. Is this compromise fine? 

The downside of this approach is that if we have to list all the collections required for all the tests in the package (for now, it is only dialout_test which requires Gateways. So not a problem).
